### PR TITLE
Bug 1424335 - used parameterized roles for gecko branches

### PR DESCRIPTION
**Pull request by djmitche**
This uses the project as the parameter, and then replicates the roles
for levels and trust domains.  Believe me, supporting multiple
parameters was just not possible :)

### CODE REVIEW

**Comment by jonasfj** [comment]
I'm not completely following, I'll get back to this again later..

The only comment I have is a light-weight description of terms like:
 * feature
 * trust-domain
 * project

and how they intersect...

File: src/make-gecko-branch-role.js [hunk 1]
```diff
@@ -48,64 +55,13 @@ module.exports.run = async function(projectsOption, options) {
...
-        'index:insert-task:buildbot.revisions.*',
-        'project:releng:buildbot-bridge:builder-name:release-<project>-*',
-        'project:releng:buildbot-bridge:builder-name:release-<project>_*',
-      ].map((scope) =>
-        scope
-        .replace('<project>', projectName)
-        .replace('<level>', level)
-      ));
-    }
+      `assume:project:releng:branch:${domain}:level-${level}:${projectName}`,
```
Comment (tomprince) [inline src/make-gecko-branch-role.js]: Does anything already have this scope, or a prefix of it? I know there are scopes under `project:releng:` already.

Also, thunderbird has [started](https://bugzilla.mozilla.org/show_bug.cgi?id=1424254) using `project:comm:thunderbird:releng` instead of `project:releng`.

File: src/make-gecko-parameterized-roles.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+          `assume:moz-tree:level:${level}:${domain}`,
+          `queue:route:index.${domain}.v2.<..>.*`,
+          `index:insert-task:${domain}.v2.<..>.*`,
+          `queue:route:index.${domain}.cache.level-${level}.*`,
+          `index:insert-task:${domain}.cache.level-${level}.*`,
+          `queue:route:tc-treeherder-stage.<..>.*`,
+          `queue:route:tc-treeherder.<..>.*`,
+          `queue:route:tc-treeherder-stage.v2.<..>.*`,
+          `queue:route:tc-treeherder.v2.<..>.*`,
+          `queue:route:coalesce.v1.builds.<..>.*`,  // deprecated - https://bugzilla.mozilla.org/show_bug.cgi?id=1382204
```
Comment (tomprince) [inline src/make-gecko-parameterized-roles.js:48]: This and the tc-treeherder-stage routes are old, and could perhaps be moved to a deprecated feature? Although, I guess maybe that wants to be done separately.

**Review by tomprince — CHANGES_REQUESTED** [review]
I like the direction of this. I'm not sure about the names of the new parameterized scopes, (the `feature` bit in particular, bothers me, although I can't articulate why).

I think it would be good to have some documentation on what is going on:
- some prose documentation explaining what the roles are and how https://hg.mozilla.org/build/ci-configuration/ comes into play
- some documentation for each of the commands (docstrings or prose or something)
- documentation of each of the "features"

I'm not sure how much of that documentation should live here or elsewhere.

**Review by jonasfj — APPROVED** [review]
From a quick look this looks sane...

But I would really prefer a much more declarative approach.
Declare the roles or role patterns you want to created and write them in YAML or JSON... or YAML + jsone..

But the mix of code and config is hard to follow...

**Comment by djmitche** [comment]
Well, this is all moving in-tree and into Python at some point, at which time it will be fairly declarative.
